File: public/code/memecoin__bonding_curve_sniper.ts
// Talons Strategies - Bonding Curve Sniper (Pump.fun / BonkFun / Bags.fm)
import { makeSignal, Signal } from "../../common";

export interface BondingCurveData {
  symbol: string;
  bonding_curve_pct: number;
  holders: number;
  buys_5m: number;
  sells_5m: number;
  volume_usd_5m: number;
  price_usd: number;
  age_minutes: number;
}

export interface BCSConfig {
  min_holders?: number;
  min_volume_usd_5m?: number;
  min_buy_ratio?: number;
  min_age_minutes?: number;
  max_age_minutes?: number;
  sweet_zone_low?: number;
  sweet_zone_high?: number;
  exit_curve_pct?: number;
}

export function checkSignal(data: BondingCurveData, cfg: BCSConfig = {}): Signal {
  const minH  = cfg.min_holders ?? 50;
  const minV  = cfg.min_volume_usd_5m ?? 5000;
  const minBR = cfg.min_buy_ratio ?? 1.5;
  const minA  = cfg.min_age_minutes ?? 2;
  const maxA  = cfg.max_age_minutes ?? 240;
  const szL   = cfg.sweet_zone_low ?? 20;
  const szH   = cfg.sweet_zone_high ?? 60;
  const exitP = cfg.exit_curve_pct ?? 90;

  const { symbol, bonding_curve_pct: curve, holders, buys_5m: buys, sells_5m: sells,
          volume_usd_5m: vol, price_usd: price, age_minutes: age } = data;

  if (age < minA) return makeSignal("neutral", price, 0, [], 0, `Too fresh (${age}min)`);
  if (age > maxA) return makeSignal("neutral", price, 0, [], 0, `Too old (${age}min)`);
  if (curve >= exitP)
    return makeSignal("short", price, price*1.15, [price*0.85, price*0.7, price*0.5], 0.7,
                     `Pre-migration exit (${curve.toFixed(1)}%)`);
  if (holders < minH) return makeSignal("neutral", price, 0, [], 0, `Low holders (${holders})`);
  if (vol < minV)     return makeSignal("neutral", price, 0, [], 0, `Low volume ($${vol.toFixed(0)})`);

  const buyRatio = buys / Math.max(sells, 1);
  if (buyRatio < minBR)
    return makeSignal("neutral", price, 0, [], 0, `Weak buy ratio (${buyRatio.toFixed(2)})`);

  const inSweet = curve >= szL && curve <= szH;
  let conf = 0.5;
  if (inSweet)         conf += 0.2;
  if (buyRatio > 2.5)  conf += 0.1;
  if (holders > 200)   conf += 0.1;
  if (vol > 20000)     conf += 0.1;
  conf = Math.min(conf, 0.95);

  return makeSignal("long", price, price*0.7, [price*1.5, price*2.5, price*4.0], conf,
                    `BCS: curve=${curve.toFixed(1)}% holders=${holders} buyR=${buyRatio.toFixed(2)} vol=$${vol.toFixed(0)}`,
                    { symbol, zone: inSweet ? "sweet" : curve < szL ? "early" : "late" });
}

File: public/code/memecoin__bonding_curve_sniper.py
"""
Talons Strategies - Bonding Curve Sniper (Pump.fun / BonkFun / Bags.fm)

STRATEJI:
- Yeni token launch'larında bonding curve dolum yüzdesine göre sinyal üret
- %0-20: Risky/early, en yüksek upside
- %20-60: Sweet spot - momentum kanıtlanmış, hala curve'de
- %60-90: Pre-migration zone, dikkatli
- %90+: Migration imminent, çıkış zamanı
- Volume + holder growth + buy/sell ratio filtreler ekleniyor

INPUT (off-chain data):
{
    "symbol": "TOKEN",
    "bonding_curve_pct": 0-100,   # curve doluluk %
    "holders": int,
    "buys_5m": int,
    "sells_5m": int,
    "volume_usd_5m": float,
    "price_usd": float,
    "age_minutes": int,
}
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", ".."))
from common import make_signal


def check_signal(data: dict, cfg: dict = None) -> dict:
    cfg = cfg or {}
    min_holders   = cfg.get("min_holders", 50)
    min_volume    = cfg.get("min_volume_usd_5m", 5000)
    min_buy_ratio = cfg.get("min_buy_ratio", 1.5)
    min_age       = cfg.get("min_age_minutes", 2)
    max_age       = cfg.get("max_age_minutes", 240)
    sweet_low     = cfg.get("sweet_zone_low", 20)
    sweet_high    = cfg.get("sweet_zone_high", 60)
    exit_pct      = cfg.get("exit_curve_pct", 90)

    symbol = data.get("symbol", "?")
    curve  = float(data.get("bonding_curve_pct", 0))
    holders= int(data.get("holders", 0))
    buys   = int(data.get("buys_5m", 0))
    sells  = int(data.get("sells_5m", 1))
    vol    = float(data.get("volume_usd_5m", 0))
    price  = float(data.get("price_usd", 0))
    age    = int(data.get("age_minutes", 0))

    if age < min_age:
        return make_signal("neutral", price, 0, [], 0, f"Too fresh ({age}min)")
    if age > max_age:
        return make_signal("neutral", price, 0, [], 0, f"Too old ({age}min)")
    if curve >= exit_pct:
        return make_signal("short", price, price * 1.15, [price*0.85, price*0.7, price*0.5], 0.7,
                           f"Pre-migration exit zone ({curve:.1f}%)")
    if holders < min_holders:
        return make_signal("neutral", price, 0, [], 0, f"Low holders ({holders})")
    if vol < min_volume:
        return make_signal("neutral", price, 0, [], 0, f"Low volume (${vol:.0f})")

    buy_ratio = buys / max(sells, 1)
    if buy_ratio < min_buy_ratio:
        return make_signal("neutral", price, 0, [], 0, f"Weak buy ratio ({buy_ratio:.2f})")

    in_sweet = sweet_low <= curve <= sweet_high
    conf = 0.5
    if in_sweet:           conf += 0.2
    if buy_ratio > 2.5:    conf += 0.1
    if holders > 200:      conf += 0.1
    if vol > 20000:        conf += 0.1
    conf = min(conf, 0.95)

    return make_signal(
        "long", price,
        price * 0.7,
        [price * 1.5, price * 2.5, price * 4.0],
        conf,
        f"BCS: curve={curve:.1f}% holders={holders} buy_ratio={buy_ratio:.2f} vol=${vol:.0f}",
        metadata={"symbol": symbol, "zone": "sweet" if in_sweet else "early" if curve < sweet_low else "late"},
    )
